src/zkevm_specs/evm/execution/log.py

```diff
@@ -0,0 +1,60 @@
+from ..instruction import Instruction, Transition
+from ..table import CallContextFieldTag, TxLogFieldTag, TxContextFieldTag
+from ..opcode import Opcode
+from ...util.param import LOG_STATIC_GAS
+
+
+def log(instruction: Instruction):
+    opcode = instruction.opcode_lookup(True)
+    # constrain op in [log0, log4] range
+    instruction.range_lookup(opcode - Opcode.LOG0, 5)
+
+    # pop `mstart`, `msize` from stack
+    mstart = instruction.rlc_to_int_exact(instruction.stack_pop(), 8)
+    msize = instruction.rlc_to_int_exact(instruction.stack_pop(), 8)
+    topics = []
+    memory_data = []
+    for i in range(opcode - Opcode.LOG0):
+        stack_topic = instruction.stack_pop()
+        topics.append(stack_topic)
+
+    # check memory copy
+    for i in range(msize):
+        address = mstart + i + 1
+        data_byte = instruction.memory_read(address, instruction.curr.call_id)
+        memory_data.append(data_byte)
+
+    # constrain topics in logs
+    for i in range(len(topics)):
+        topic_in_log = instruction.tx_log_lookup(TxLogFieldTag.Topics, i)
+        instruction.constrain_equal(topic_in_log, topics[i])
+
+    # constrain data in logs
+    for i in range(len(memory_data)):
+        byte_in_log = instruction.tx_log_lookup(TxLogFieldTag.Data, i)
+        instruction.constrain_equal(byte_in_log, memory_data[i])
+
+    # check not static call
+    instruction.constrain_equal(0, instruction.call_context_lookup(CallContextFieldTag.IsStatic))
+
+    # check contract address validity, use call context's  callee address as contract address
+    contract_address = instruction.call_context_lookup(CallContextFieldTag.CalleeAddress)
+    instruction.constrain_equal(contract_address, instruction.tx_log_lookup(TxLogFieldTag.Address))
+
+    # omit block number constraint even it is set within op code explicitly, because by default the circuit only handle
+    # current block, otherwise, block context lookup is required.
+    # calculate dynamic gas cost
+    next_memory_size, memory_expansion_cost = instruction.memory_expansion_constant_length(mstart, msize)
+    dynamic_gas = LOG_STATIC_GAS * (opcode - Opcode.LOG0) + 8 * msize + memory_expansion_cost
+
+    rw_counter_diff = 5 + 2 * (opcode - Opcode.LOG0) + 2 * msize
```

Comment:
Can we add some comments to explain the number?

Reply:
adding in specs/opcode/A0_A4LOG.md